src/gui/qgsmapcanvas.cpp

```diff
@@ -3627,3 +3629,37 @@ double QgsMapCanvas::zoomOutFactor() const
   }
   return mWheelZoomFactor;
 }
+
+QgsRectangle QgsMapCanvas::computeExtentForRotation( const QgsRectangle &extent, double rotation ) const
+{
+  if ( extent.isEmpty() || qgsDoubleNear( rotation, 0.0 ) )
+  {
+    return extent;
+  }
+  // Compute the oriented bounding box of the extent
+  QgsGeometry geom = QgsGeometry::fromRect( extent );
+  geom.rotate( rotation, geom.centroid().asPoint() );
+  const QgsRectangle oob = geom.boundingBox();
```

Comment:
This could be made considerably more efficient -- eg by using something like https://codereview.stackexchange.com/a/267876 (maybe as a method in QgsGeometryUtils, or in QgsRectangle directly)

The current approach involves GEOS to determine the centroid, which means we have conversions from QgsGeometry -> GEOS -> back involved.

Reply:
Yeah you're right, we just need a simple rect rotation, so we shouldn't call GEOS at all. I'll fix it.